processing/src/main/java/org/apache/druid/segment/filter/Filters.java

```diff
@@ -552,4 +594,96 @@ public static boolean filterMatchesNull(Filter filter)
     ValueMatcher valueMatcher = filter.makeMatcher(ALL_NULL_COLUMN_SELECTOR_FACTORY);
     return valueMatcher.matches();
   }
+
+  /**
+   * Flattens children of an AND, removes duplicates, and removes literally-true filters.
+   */
+  private static List<Filter> flattenAndChildren(final List<Filter> filters)
+  {
+    final List<Filter> retVal = new ArrayList<>();
+    final Set<EquivalenceCheckedFilter> seenFilters = new HashSet<>();
+
+    for (Filter child : filters) {
+      if (child instanceof AndFilter) {
+        retVal.addAll(flattenAndChildren(((AndFilter) child).getFilters()));
+      } else if (!(child instanceof TrueFilter) && seenFilters.add(new EquivalenceCheckedFilter(child))) {
+        retVal.add(child);
+      }
+    }
+
+    return retVal;
+  }
+
+  /**
+   * Flattens children of an OR, removes duplicates, and removes literally-false filters.
+   */
+  private static List<Filter> flattenOrChildren(final List<Filter> filters)
+  {
+    final List<Filter> retVal = new ArrayList<>();
+    final Set<EquivalenceCheckedFilter> seenFilters = new HashSet<>();
+
+    for (Filter child : filters) {
+      if (child instanceof OrFilter) {
+        retVal.addAll(flattenOrChildren(((OrFilter) child).getFilters()));
+      } else if (!(child instanceof FalseFilter) && seenFilters.add(new EquivalenceCheckedFilter(child))) {
+        retVal.add(child);
+      }
+    }
+
+    return retVal;
+  }
+
+  /**
+   * Wraps a {@link Filter} in an object whose {@link #equals} and {@link #hashCode} methods are overridden to
+   * ignore order of children for {@link AndFilter} and {@link OrFilter}. Useful for deduplication in
+   * {@link #flattenAndChildren} and {@link #flattenOrChildren}.
+   */
+  private static class EquivalenceCheckedFilter
+  {
+    private final Filter filter;
+
+    EquivalenceCheckedFilter(Filter filter)
+    {
+      this.filter = filter;
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      EquivalenceCheckedFilter that = (EquivalenceCheckedFilter) o;
+
+      if (filter instanceof BooleanFilter && filter.getClass().equals(that.filter.getClass())) {
+        // Order of children doesn't matter for equivalence.
+        final Set<Filter> ourFilterSet = new HashSet<>(((BooleanFilter) filter).getFilters());
+        final Set<Filter> theirFilterSet = new HashSet<>(((BooleanFilter) that.filter).getFilters());
+        return ourFilterSet.equals(theirFilterSet);
```

Comment:
Does this handle nested BooleanFilters correctly?

Reply:
It handles them correctly in that it's not wrong. But, it isn't necessarily doing the best possible optimization. I edited it to use EquivalenceCheckedFilter recursively and added some more tests (see FiltersTest for the new ones).

Reply:
Nevermind, I decided to take your advice and switch the List to a LinkedHashSet, which makes this EquivalenceCheckedFilter technique no longer necessary.